doc/lua_api.md

```diff
@@ -2562,6 +2562,638 @@ print(ItemStack("mod:item_with_no_desc"):get_description()) --> mod:item_with_no
+  window, and showing the new window to the player immediately.
+    * `id`: The window ID of the window to reshow.
+    * Returns the new window ID for the new window. The old window ID is no
+      longer valid.
+* `ui.close(id)`: Closes a window that is currently shown to the player.
+    * `id`: The window ID of the window to close. This ID is no longer valid
+      after this call returns.
+* `ui.window_info(id)`: Returns a table containing information about a
+  currently open window.
+    * `id`: The window ID of the window to get information about.
+    * Returns a table with the following properties:
+        * `builder`: The builder function for this window.
+        * `player`: The player that the window is shown to.
+        * `context`: The context table for this window.
+
+Utility functions
+-----------------
+
+* `ui.new_id()`: Returns a new globally unique ID string.
+    - The format of this ID is not specified, but it will have the format of an
+      engine reserved ID and will not conflict with any other ID created during
+      this session.
+* `ui.is_id(str)`: Checks whether the argument is a string that follows the
+  format of an ID string.
+* `ui.get_default_theme()`: Returns the style used as the default theme when no
+  explicit theme is provided for a window.
+* `ui.set_default_theme(theme)`: Sets the default theme to a new style,
+  overriding the default engine-provided theme.
+
+`ui.Window`
+-----------
+
+Windows represent discrete self-contained UIs formed by a tree of elements and
+other parameters affecting the entire window.
+
+The window contains a single element, which is the root element of the element
+tree. Since the root element has no parent, it is positioned relative to the
+entire screen.
+
+### Window types
+
+Windows have a window type, which determines whether they can receive user
+input, what type of scaling to apply to the pixels, and the Z order of how they
+are drawn in relation to other things on the screen. These are the following:
+
+* `bg`: Used for things that need to be drawn before everything else.
+* `mask`: Used for visual effects covering the player's face, such as masks.
+* `hud`: Used for normal HUD purposes. Hidden when the HUD is hidden.
+* `message`: Used to display GUI-like popup messages, but that can't be
+  interacted with. Hidden when the chat is hidden.
+* `gui`: Used for GUIs that the user can interact with. If there are no
+  formspecs open but one or more `gui` windows are open, then the last opened
+  window receives user input.
+* `fg`: Used for things that need to be drawn after everything else.
+
+Only the `gui` layer can receive user input from the mouse or keyboard. No
+other layer will receive events. The `gui` and `message` window types use
+`gui_scaling` for the base pixel size, whereas every other type uses
+`hud_scaling`.
+
+**NOTE**: Until events are implemented, the `gui` layer cannot receive user
+input and is effectively a HUD layer.
```

Comment:
Regarding "only the `gui` layer can receive user input": What about touchscreens? Using a touchscreen, you can "click" HUD buttons as well. The touch controls themselves are basically a HUD.

Reply:
That's a very good point.  But we don't want such things showing up on desktop.  Maybe a separate layer that only shows up on touchscreen devices.

Reply:
> In general, I'm not going to worry too much about supporting touchscreen devices yet

I would like to avoid ending up in a similar situations as with formspecs, which are terrible to use on mobile (way  too small, no native gestures like "swipe to scroll", etc) ([nice comment by rollerozxa](https://github.com/Bithack/principia/issues/95#issue-1618844353)). It would be great if mobile could become a first-class platform. But worrying about it won't help, of course :)

Reply:
Well, in the sense that I don't know how to compile Minetest for touchscreen devices and don't want to mess with Android Studio, I'm not planning on supporting them yet.

However, if we get SDL running, then we have native touchscreen events on every platform.  I do have a laptop with a touchscreen (laptops _without_ touchscreens seem increasingly rare...), so I would definitely add touch support then.  But that's only after SDL is ready.

Reply:
> However, if we get SDL running, then we have native touchscreen events on every platform. I do have a laptop with a touchscreen (laptops without touchscreens seem increasingly rare...), so I would definitely add touch support then. But that's only after SDL is ready.

That's already done since two days ago: minetest/irrlicht#209 and minetest/irrlicht#262 are merged. You can test on your laptop :)